<?xml version="1.0" encoding="ascii"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
          "DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>Scientific.IO.PDB</title>
  <link rel="stylesheet" href="epydoc.css" type="text/css" />
  <script type="text/javascript" src="epydoc.js"></script>
</head>

<body bgcolor="white" text="black" link="blue" vlink="#204080"
      alink="#204080">
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th class="navbar">&nbsp;&nbsp;&nbsp;<a class="navbar"
        href="Scientific-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th class="navbar">&nbsp;&nbsp;&nbsp;<a class="navbar"
        href="trees.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th class="navbar">&nbsp;&nbsp;&nbsp;<a class="navbar"
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th class="navbar">&nbsp;&nbsp;&nbsp;<a class="navbar"
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center">
            <p class="nomargin">
              <a class="navbar" target="_top" href="http://dirac.cnrs-orleans.fr/ScientificPython/">Scientific Python</a>
      </p></th></tr></table></th>
  </tr>
</table>
<table width="100%" cellpadding="0" cellspacing="0">
  <tr valign="top">
    <td width="100%">
      <span class="breadcrumbs">
        <a href="Scientific-module.html">Package&nbsp;Scientific</a> ::
        <a href="Scientific.IO-module.html">Package&nbsp;IO</a> ::
        Module&nbsp;PDB
      </span>
    </td>
    <td>
      <table cellpadding="0" cellspacing="0">
        <!-- hide/show private -->
        <tr><td align="right"><span class="options">[<a href="javascript: void(0);" class="privatelink"
    onclick="toggle_private();">hide private</a>]</span></td></tr>
        <tr><td align="right"><span class="options"
            >[<a href="frames.html" target="_top">frames</a
            >]&nbsp;|&nbsp;<a href="Scientific.IO.PDB-module.html"
            target="_top">no&nbsp;frames</a>]</span></td></tr>
      </table>
    </td>
  </tr>
</table>
<!-- ==================== MODULE DESCRIPTION ==================== -->
<h2 class="module">Module PDB
<br/><span class="codelink"><a href="Scientific.IO.PDB-pysrc.html">source&nbsp;code</a></span></h2>
<p>Parsing and writing of Protein Data Bank (PDB) files</p>
  <p>This module provides classes that represent PDB (Protein Data Bank) 
  files and configurations contained in PDB files. It provides access to 
  PDB files on two levels: low-level (line by line) and high-level (chains,
  residues, and atoms).</p>
  <p>Caution: The PDB file format has been heavily abused, and it is 
  probably impossible to write code that can deal with all variants 
  correctly. This modules tries to read the widest possible range of PDB 
  files, but gives priority to a correct interpretation of the PDB format 
  as defined by the Brookhaven National Laboratory.</p>
  <p>A special problem are atom names. The PDB file format specifies that 
  the first two letters contain the right-justified chemical element name. 
  A later modification allowed the initial space in hydrogen names to be 
  replaced by a digit. Many programs ignore all this and treat the name as 
  an arbitrary left-justified four-character name. This makes it difficult 
  to extract the chemical element accurately; most programs write the 
  '&quot;CA&quot;' for C_alpha in such a way that it actually stands for a 
  calcium atom. For this reason a special element field has been added 
  later, but only few files use it. In the absence of an element field, the
  code in this module attempts to guess the element using all information 
  available.</p>
  <p>The low-level routines in this module do not try to deal with the atom
  name problem; they return and expect four-character atom names including 
  spaces in the correct positions. The high-level routines use atom names 
  without leading or trailing spaces, but provide and use the element field
  whenever possible. For output, they use the element field to place the 
  atom name correctly, and for input, they construct the element field 
  content from the atom name if no explicit element field is found in the 
  file.</p>
  <p>Except where indicated, numerical values use the same units and 
  conventions as specified in the PDB format description.</p>
  Example:
<pre class="literalblock">
 &gt;&gt;&gt;conf = Structure('example.pdb')
 &gt;&gt;&gt;print conf
 &gt;&gt;&gt;for residue in conf.residues:
 &gt;&gt;&gt;    for atom in residue:
 &gt;&gt;&gt;        print atom
</pre><br /><br />

<!-- ==================== CLASSES ==================== -->
<a name="section-Classes"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="summary">
  <td colspan="2" class="summary">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><h3 class="summary">Classes</h3></th>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Classes"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB.AminoAcidResidue-class.html">AminoAcidResidue</a></strong></td>
  <td class="summary">Amino acid residue in a PDB file</td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB.Atom-class.html">Atom</a></strong></td>
  <td class="summary">Atom in a PDB structure</td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB.Chain-class.html">Chain</a></strong></td>
  <td class="summary">Chain of PDB residues</td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB.Group-class.html">Group</a></strong></td>
  <td class="summary">Atom group (residue or molecule) in a PDB file</td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB.HetAtom-class.html">HetAtom</a></strong></td>
  <td class="summary">HetAtom in a PDB structure</td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB.Molecule-class.html">Molecule</a></strong></td>
  <td class="summary">Molecule in a PDB file</td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB.NucleotideChain-class.html">NucleotideChain</a></strong></td>
  <td class="summary">Nucleotide chain in a PDB file</td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB.NucleotideResidue-class.html">NucleotideResidue</a></strong></td>
  <td class="summary">Nucleotide residue in a PDB file</td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB.PDBFile-class.html">PDBFile</a></strong></td>
  <td class="summary"><a name="index-PDB"></a><i class="indexterm">PDB</i> file with 
        access at the record level</td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB.PeptideChain-class.html">PeptideChain</a></strong></td>
  <td class="summary">Peptide chain in a PDB file</td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB.Residue-class.html">Residue</a></strong></td>
  <td class="summary">&nbsp;</td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB.ResidueNumber-class.html">ResidueNumber</a></strong></td>
  <td class="summary">PDB residue number</td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB.Structure-class.html">Structure</a></strong></td>
  <td class="summary">A high-level representation of the contents of a PDB file</td>
</tr>
</table>

<br />
<!-- ==================== FUNCTIONS ==================== -->
<a name="section-Functions"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="summary">
  <td colspan="2" class="summary">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><h3 class="summary">Functions</h3></th>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Functions"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
  <td width="15%" align="right" valign="top" class="summary">
    <span class="rtype">
      &nbsp;
    </span>
   </td>
  <td class="summary">
    <span class="sig"><a href="Scientific.IO.PDB-module.html#defineAminoAcidResidue" class="sig-name">defineAminoAcidResidue</a>(<span class="sig-arg">symbol</span>)</span>
    <br />Make the parser recognize a particular residue type as an amino 
      acid residue
  </td>
</tr>
<tr>
  <td width="15%" align="right" valign="top" class="summary">
    <span class="rtype">
      &nbsp;
    </span>
   </td>
  <td class="summary">
    <span class="sig"><a href="Scientific.IO.PDB-module.html#defineNucleicAcidResidue" class="sig-name">defineNucleicAcidResidue</a>(<span class="sig-arg">symbol</span>)</span>
    <br />Make the parser recognize a particular residue type as an nucleic 
      acid residue
  </td>
</tr>
</table>

<br />
<!-- ==================== VARIABLES ==================== -->
<a name="section-Variables"></a>
<table class="summary" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="summary">
  <td colspan="2" class="summary">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><h3 class="summary">Variables</h3></th>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-Variables"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB-module.html#amino_acids">amino_acids</a></strong></td>
  <td class="summary">&nbsp;</td>
</tr>
<tr>
  <td width="15%" class="summary">
    <strong><a href="Scientific.IO.PDB-module.html#nucleic_acids">nucleic_acids</a></strong></td>
  <td class="summary">&nbsp;</td>
</tr>
</table>

<br />
<!-- ==================== FUNCTION DETAILS ==================== -->
<a name="section-FunctionDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="details">
  <td colspan="2" class="details">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><h3 class="details">Function Details</h3></th>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-FunctionDetails"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
</table>
<a name="defineAminoAcidResidue"></a>
<div>
<table width="100%" class="details" bgcolor="#e0e0e0"><tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3><span class="sig"><span class="sig-name">defineAminoAcidResidue</span>(<span class="sig-arg">symbol</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="Scientific.IO.PDB-pysrc.html#defineAminoAcidResidue">source&nbsp;code</a></span>&nbsp;</span
    ></td>
  </table>
  
  Make the parser recognize a particular residue type as an amino acid 
  residue
  <dl><dt></dt><dd>
    <dl><dt>Parameters:</dt></dl>
    <ul>
        <li><strong class="pname"><code>symbol</code></strong> (<code><code>string</code></code>) - the three-letter code for an amino acid</li>
    </ul>
  </dd></dl>
</td></tr></table>
</div>
<a name="defineNucleicAcidResidue"></a>
<div>
<table width="100%" class="details" bgcolor="#e0e0e0"><tr><td>
  <table width="100%" cellpadding="0" cellspacing="0" border="0">
  <tr valign="top"><td>
  <h3><span class="sig"><span class="sig-name">defineNucleicAcidResidue</span>(<span class="sig-arg">symbol</span>)</span>
  </h3>
  </td><td align="right" valign="top"
    ><span class="codelink"><a href="Scientific.IO.PDB-pysrc.html#defineNucleicAcidResidue">source&nbsp;code</a></span>&nbsp;</span
    ></td>
  </table>
  
  Make the parser recognize a particular residue type as an nucleic acid
  residue
  <dl><dt></dt><dd>
    <dl><dt>Parameters:</dt></dl>
    <ul>
        <li><strong class="pname"><code>symbol</code></strong> (<code><code>string</code></code>) - the one-letter code for a nucleic acid</li>
    </ul>
  </dd></dl>
</td></tr></table>
</div>
<br />
<!-- ==================== VARIABLES DETAILS ==================== -->
<a name="section-VariablesDetails"></a>
<table class="details" border="1" cellpadding="3"
       cellspacing="0" width="100%" bgcolor="white">
<tr bgcolor="#70b0f0" class="details">
  <td colspan="2" class="details">
    <table border="0" cellpadding="0" cellspacing="0" width="100%">
      <tr valign="top">
        <td align="left"><h3 class="details">Variables Details</h3></th>
        <td align="right" valign="top"
         ><span class="options">[<a href="#section-VariablesDetails"
         class="privatelink" onclick="toggle_private();"
         >hide private</a>]</span></td>
      </tr>
    </table>
  </td>
</tr>
</table>
<a name="amino_acids"></a>
<div>
<table width="100%" class="details" bgcolor="#e0e0e0"><tr><td>
  <h3>amino_acids</h3>
  
  <dl><dt></dt><dd>
    <dl title="['ALA', 'ARG', 'ASN', 'ASP', 'CYS', 'CYX', 'GLN', 'GLU', 'GLY', 'HIS', 'HID', 'HIE', 'HIP', 'HSD', 'HSE', 'HSP', 'ILE', 'LEU', 'LYS', 'MET', 'PHE', 'PRO', 'SER', 'THR', 'TRP', 'TYR', 'VAL', 'ACE', 'NME', 'NHE']"><dt>Value:</dt>
      <dd><table><tr><td><pre class="variable">
['ALA', 'ARG', 'ASN', 'ASP', 'CYS', 'CYX', 'GLN', 'GLU', 'GLY']        
      </pre></td></tr></table></dd>
    </dl>
  </dd></dl>
</td></tr></table>
</div>
<a name="nucleic_acids"></a>
<div>
<table width="100%" class="details" bgcolor="#e0e0e0"><tr><td>
  <h3>nucleic_acids</h3>
  
  <dl><dt></dt><dd>
    <dl title="['A', 'C', 'G', 'I', 'T', 'U', '+A', '+C', '+G', '+I', '+T', '+U', 'RA', 'RC', 'RG', 'RU', 'DA', 'DC', 'DG', 'DT', 'RA5', 'RC5', 'RG5', 'RU5', 'DA5', 'DC5', 'DG5', 'DT5', 'RA3', 'RC3', 'RG3', 'RU3', 'DA3', 'DC3', 'DG3', 'DT3', 'RAN', 'RCN', 'RGN', 'RUN', 'DAN', 'DCN', 'DGN', 'DTN']"><dt>Value:</dt>
      <dd><table><tr><td><pre class="variable">
['A', 'C', 'G', 'I', 'T', 'U', '+A', '+C', '+G']                       
      </pre></td></tr></table></dd>
    </dl>
  </dd></dl>
</td></tr></table>
</div>
<br />
<!-- ==================== NAVIGATION BAR ==================== -->
<table class="navbar" border="0" width="100%" cellpadding="0"
       bgcolor="#a0c0ff" cellspacing="0">
  <tr valign="middle">
  <!-- Home link -->
      <th class="navbar">&nbsp;&nbsp;&nbsp;<a class="navbar"
        href="Scientific-module.html">Home</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Tree link -->
      <th class="navbar">&nbsp;&nbsp;&nbsp;<a class="navbar"
        href="trees.html">Trees</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Index link -->
      <th class="navbar">&nbsp;&nbsp;&nbsp;<a class="navbar"
        href="identifier-index.html">Indices</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Help link -->
      <th class="navbar">&nbsp;&nbsp;&nbsp;<a class="navbar"
        href="help.html">Help</a>&nbsp;&nbsp;&nbsp;</th>

  <!-- Project homepage -->
      <th class="navbar" align="right" width="100%">
        <table border="0" cellpadding="0" cellspacing="0">
          <tr><th class="navbar" align="center">
            <p class="nomargin">
              <a class="navbar" target="_top" href="http://dirac.cnrs-orleans.fr/ScientificPython/">Scientific Python</a>
      </p></th></tr></table></th>
  </tr>
</table>
<table border="0" cellpadding="0" cellspacing="0" width="100%%">
  <tr>
    <td align="left" class="footer">Generated by Epydoc
        3.0alpha2 on Tue Jun 13 11:39:41 2006</td>
    <td align="right" class="footer">
      <a href="http://epydoc.sourceforge.net">http://epydoc.sf.net</a>
    </td>
  </tr>
</table>

<script type="text/javascript">
  <!--
  // Private objects are initially displayed (because if
  // javascript is turned off then we want them to be
  // visible); but by default, we want to hide them.  So hide
  // them unless we have a cookie that says to show them.
  checkCookie()
  // -->
</script>
  
</body>
</html>
